Split out MySqlDataReader ResultSet and Row operations into classes

caleblloyd opened the pull request:
- Creates a `MySql.Data.MySqlClient.Results` namespace
  - `ResultSet` class holds Result Set oriented behavior
  - `Row` class holds Row oriented behavior
- Allows for Buffering Result Set or Row in order to look ahead where needed, such as in #135 
- Makes `MySqlDataReader` a lot easier to comprehend :smile: 
- Opening PR for design feedback and suggestions

bgrainger commented on src/MySqlConnector/MySqlClient/MySqlDataReader.cs:
```diff
@@ -628,205 +165,77 @@ protected override void Dispose(bool disposing)
-			NoMoreData,
+			get { return _resultSet;  }
+			set { _resultSet = value?.Actiate() ?? new ResultSet(this); }
 		}
```
From an API design perspective, it's somewhat confusing to have `_resultSet` field, `m_ResultSet` property and `GetResultSet` method that all do something slightly different. Also, a setter that transforms `m_ResultSet = null;` into a non-null value smells like a bit of a hack that would be confusing to understand.

Perhaps the property should be removed and its setter become an explicit `SetAndActivateResultSet` method that manipulates the field?

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,358 @@
+			return this;
+		}
+
+		internal ResultSet Actiate()
```
Typo: `Actiate`

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,358 @@
+			DataReader = dataReader;
+		}
+
+		internal async Task<ResultSet> ReadResultSetHeaderAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
```
Nitpick: I would consider this member to be part of the "public" API of the class (irrespective of the class's `internal` accessibility), following Eric's arguments here: https://ericlippert.com/2014/09/15/internal-or-public/

(Most of the methods on this type I would make `public`.)

bgrainger replied:
I prefer `ValueTask<T>` (fewer allocations and async state machines on the happy path) to `Task` for APIs that aren't required by ADO.NET.

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,358 @@
+
+		internal async Task<Row> BufferReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			var read = await DoReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
```
`row` would be a better name.

bgrainger commented:
It seems like this might be doing a lot more allocations than the previous code, but (without measuring) I don't know if it's significant enough to impact performance when reading large resultsets.

bgrainger commented on src/MySqlConnector/MySqlClient/MySqlDataReader.cs:
```diff
@@ -16,578 +16,105 @@ public sealed class MySqlDataReader : DbDataReader
 		}
 
-		public override string GetDataTypeName(int ordinal)
+		internal async Task<ResultSet> DoNextResultAsync(IOBehavior ioBehavior, ResultSet currentResultSet, CancellationToken cancellationToken)
```
`private`

caleblloyd commented:
Allocations should be limited to a few pointers per result set and a few pointers per row if no buffering is used (the default).  I'm trying to get the buffering working reliably right now.  I'll see if I can make it re-use objects when no buffering is used to cut down on allocations.

bgrainger commented on src/MySqlConnector/MySqlClient/MySqlDataReader.cs:
```diff
@@ -16,578 +16,105 @@ public sealed class MySqlDataReader : DbDataReader
-			VerifyHasResult();
-
-			for (int column = 0; column < m_columnDefinitions.Length; column++)
+			catch (Exception)
```
`Exception` is a broad type to catch here; are there specific exceptions that should be getting handled?

bgrainger commented:
I ran the following code under VS' .NET Memory Allocation profiler:

```
await connection.OpenAsync();

using (var cmd = new MySqlCommand(@"drop table if exists integers; create table integers(value integer not null primary key); insert into integers(value) values(0), (1), (2), (3), (4), (5), (6), (7), (8), (9);", connection))
	await cmd.ExecuteNonQueryAsync();

long sum = 0;
using (var cmd = new MySqlCommand(@"select i1.value, i2.value, i3.value from integers i1 join integers i2 join integers i3;", connection))
using (var reader = await cmd.ExecuteReaderAsync())
{
	while (await reader.ReadAsync())
		sum += reader.GetInt64(0) + reader.GetInt64(1) + reader.GetInt64(2);
}
Console.WriteLine(sum);
```

Current `master` results:

![image](https://cloud.githubusercontent.com/assets/188129/20288216/87cec188-aa86-11e6-8dc9-c7b5561cb613.png)

This PR results (highlighting added from manual comparison):

![image](https://cloud.githubusercontent.com/assets/188129/20288223/8f7bd754-aa86-11e6-92d1-cfd0c7b230fa.png)

There are obviously more `Row` objects being created, but this seems to be more than offset by a reduction in `Func` allocations. (I want to understand that to see if I can remove more unnecessary Func allocations in other places.)

With a switch from `Task` to `ValueTask` in this PR, we might be able to see that number come down under the baseline.

bgrainger commented on src/MySqlConnector/MySqlClient/Results/Row.cs:
```diff
@@ -0,0 +1,394 @@
+
+			var data = new ArraySegment<byte>(m_payload, ResultSet.DataOffsets[ordinal], ResultSet.DataLengths[ordinal]);
+			var columnDefinition = ResultSet.ColumnDefinitions[ordinal];
+			var isUnsigned = columnDefinition.ColumnFlags.HasFlag(ColumnFlags.Unsigned);
```
Memory profiler says there is an allocation (!) here: #137.

caleblloyd commented:
That memory allocation is very useful, is there a good reference on how to get that working?  I have VS Community 2015 installed on my Windows Partition but I couldn't find it.

I'm seeing some other interesting behavior.  While buffering in this PR does not work perfectly yet, it appears to work fine for a single result set for the performance tests.  I've done a performance test with a large result set (return the top 200 rows in the BlogPosts table).  When I buffer the entire ResultSet into memory before reading any of it, I get around 30-50% faster response times.

This is drastically more allocations (it makes heavy use of `Buffer.BlockCopy` into `ArrayPool<byte>.Shared` buffers), yet it is faster.  Maybe .NET incurs a performance penalty for jumping from socket to socket.  In #117 I predicted that it may be due to TCP Retransmissions, but I see the same behavior over Unix Domain Sockets so I'm not sure why this is the case.

bgrainger commented:
Not sure if it's in Community Edition, but I got to it through:

* create a console app that targets .NET Framework 4.6; set it as Startup Project
* Analyze > Performance Profiler
* Select Performance Wizard; click Start
* Select .NET memory allocation; Next; Next; Finish
* the app will run and close, when it does, the analysis results will be displayed automatically
* select Current View: Allocation at the top to show this list

bgrainger commented:
As "simple" as that! (I'm hoping there's a quicker way to start a new session because 10+ clicks every time will get tiresome...)

caleblloyd commented:
I think I've fixed everything in the code review, with the exception of the switch to `ValueTask`.  Most (all?) of the async code here is making or awaiting network requests, which will always result in `Task` and not `Value`.  I also find it difficult to follow the program flow when switching to continuation-style programming as opposed to async/await (I think you posted previously that ValueTask will be allowed to be used with async/await soon - can't wait!)

Allocations to `ResultSet` and `Row` should be way down now (objects are re-used when they aren't buffered).

Buffering is only used in the case of #135, when the final `NextResult` must return `false` in the case of a Stored Procedure.  I've tested it more extensively, and will post a sample of how to buffer all of the result sets in a reader.

One final good code review is needed.  I like most everything here except one part is a little confusing:
- `MySqlDataReader` contains:
  - `m_resultSet` (active result set)
  - `m_resultSetBuffered` (most recent buffered result set)
- `ResultSet` contains:
  - `m_row` (active read)
  - `m_rowBuffered` (most recent buffered read)

The references to the current and most recent buffered objects in each of the classes makes the code a little difficult to read.  I couldn't think of a better way to do it efficiently though, happy to refactor if you can.

caleblloyd commented on src/MySqlConnector/MySqlClient/MySqlDataReader.cs:
```diff
@@ -628,205 +199,71 @@ protected override void Dispose(bool disposing)
 			var dataReader = new MySqlDataReader(command, behavior);
-			await dataReader.ReadResultSetHeaderAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+			await dataReader.ReadFirstResultSetAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
 			return dataReader;
```
In order to buffer the entire result set, add `while (await BufferNextResultAsync(ioBehavior, cancellationToken).ConfigureAwait(false) != null){}` before the return

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,373 @@
+				}
+			}
+			BufferState = State;
+			//await BufferEntireAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
```
Did you mean to leave this in?

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,373 @@
+			var payloadTask = Session.ReceiveReplyAsync(ioBehavior, cancellationToken);
+			if (payloadTask.IsCompletedSuccessfully)
+				row = ReadAsyncRemainder(row, payloadTask.Result);
+			row = await ReadAsyncAwaited(row, payloadTask.AsTask()).ConfigureAwait(false);
```
This looks like a bug to assign `row` twice if the condition on line 131 is true.

caleblloyd replied:
Yes, this is not the correct behavior.  I think I can change this whole method to return a ValueTask

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,373 @@
+			return ReadAsyncRemainder(row, payload);
+		}
+
+		private Row ReadAsyncRemainder(Row row, PayloadData payload)
```
It would be more idiomatic C# to return void and take a `ref Row row` parameter; this more typically implies the variable would be reused.

bgrainger replied:
Oh, right, I forgot about that limitation.

bgrainger commented on src/MySqlConnector/MySqlClient/Results/ResultSet.cs:
```diff
@@ -0,0 +1,373 @@
+		public async Task<Row> BufferReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_rowBuffered?.BufferData();
+			m_rowBuffered = await DoReadAsync(ioBehavior, null, cancellationToken).ConfigureAwait(false);
```
Won't `m_rowBuffered` already be set by `DoReadAsync`? Unless it returns `null`, in which case `null` will be enqueued in `m_readBuffer`? Could that cause problems later?

It's confusing, and potentially error-prone, to have multiple functions responsible for setting `m_rowBuffered`. (Still trying to ensure I understand the logic properly before suggesting an alternative.)

caleblloyd commented:
I've made the following updates:

- Changed name of `DoNextResultSetAsync` to `ScanResultSetAsync` in `MySqlDataReader.cs`
  - It is now returns a `ValueTask`
  - It is solely responsible for setting `m_resultSetBuffered`

- Changed name of `DoReadAsync` to `ScanRowAsync` in `ResultSet.cs`
  - It now returns a `ValueTask`
  - It is solely responsible for setting `m_rowBuffered`

Reason for the name change it to reflect that the Result Set / Row is being "scanned" into the object that is being passed.  I've seen similar terminology in Golang's database/sql package.

I ran performance tests and saw similar numbers.  I don't think this swings performance much in one way or another, but it does give us the ability to buffer result sets / rows now.
